massa-bootstrap/src/server.rs

```diff
@@ -300,6 +318,49 @@ impl BootstrapServer {
 
         Ok(())
     }
+
+    // whether the peer IP address is banned
+    async fn is_ip_allowed(
+        &self,
+        duplex: Duplex,
+        remote_address: SocketAddr,
+        whitelist: &Option<HashSet<IpAddr>>,
+        blacklist: &Option<HashSet<IpAddr>>,
+    ) -> io::Result<(Duplex, SocketAddr)> {
+        let ip = normalize_ip(remote_address.ip());
+        // whether the peer IP address is blacklisted, send back an error message
+        if let Some(ip_list) = &blacklist && ip_list.contains(&ip) {
+            let config = self.bootstrap_config.clone();
+            let mut server = BootstrapServerBinder::new(duplex, self.keypair.clone(), config.max_bytes_read_write, config.max_bootstrap_message_size, config.thread_count, config.max_datastore_key_length, config.randomness_size_bytes, config.consensus_bootstrap_part_size);
```

Comment:
Creating a new `BootstrapServerBinder` for one message is not optimal, please take it as a function argument. This will also avoid a lot of duplicated code.

Reply:
Before calling this function, there is no instance of it. It could be not used at all, this why it's created only if needed. 
It's implemented that way in the rest of the `run`function.

Reply:
May be we should create a single instance and use it in `run` function ?

Reply:
I see, the code is still seriously duplicated though. The only difference between the two cases are the error messages, the binder creation code can be written only once and the conditional statements should only take care of the messages content.

Reply:
i'll refactor this ASAP. We could share a single instance of `BootstrapServerBinder`:

```rust
let (mut server, remote_addr) = if res_connection.is_ok() {
                        let (dplx, remote_addr) = res_connection.unwrap();
                        let server = BootstrapServerBinder::new(dplx, self.keypair.clone(), self.bootstrap_config.max_bytes_read_write, self.bootstrap_config.max_bootstrap_message_size, self.bootstrap_config.thread_count, self.bootstrap_config.max_datastore_key_length, self.bootstrap_config.randomness_size_bytes, self.bootstrap_config.consensus_bootstrap_part_size);
                        // we didn't test whether Ip is allowed
```

Reply:
done